### DISCUSSION TITLE
Question: slow `torchio.ScalarImage` compared to `pydicom.dcmread`

### DISCUSSION THREAD

**Discussion by vedal** [TorchIO-project/torchio#1005] [Q&A]
Versions:
```
python 3.8.5
torch==1.13.0+cu116
torchio==0.18.86
pydicom==2.3.1
```

Hi!
Thanks for this great package! 
I find that loading a dicom using `pydicom.dcmread` is much faster than `torchio.ScalarImage`. I tried avoiding the NaN-check. Is there any other method to speed it up?

```
%timeit -n 5 torch.tensor(pydicom.dcmread(img_path).pixel_array)
>>> 658 ms ± 9.19 ms per loop (mean ± std. dev. of 7 runs, 5 loops each)

%timeit -n 5 torchio.ScalarImage(img_path, check_nans=False).data
>>> 2.39 s ± 28.4 ms per loop (mean ± std. dev. of 7 runs, 5 loops each)
```

The dicom data has shape `(80, 758, 1016, 3)`, which `torchio.ScalarImage` automatically reshapes to `torch.Size([3, 1016, 758, 80])`.

Thanks in advance :)

**Comment by fepegar** [comment]
Hi, @vedal. Thanks for your kind words and for sharing this. We could consider switching to `pydicom` for loading DICOM. Can you please share some data to reproduce your results? This is what I'm getting:

```python
In [15]: %timeit torch.tensor(pydicom.dcmread('IMG0001.dcm').pixel_array)
1.94 ms ± 728 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [16]: %timeit torch.tensor(tio.ScalarImage('IMG0001.dcm').data)
<magic-timeit>:1: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
1.65 ms ± 484 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Reply by vedal** [reply]
Thanks for your response @fepegar.
Hm, it might depend on size of the dicom then. I'm not sure how large or how many frames your dicom was. Maybe its a thing of multi-frame dicoms? For example, the following multi-frame [dicom](http://www.chameleon-software.de/shared/examples/download/ileitis_in_yersiniosis_42_0001.dic) I found online, `shape=((42, 576, 768, 3)` seems to already show the contours of the problem:
```python
In [8]: %timeit -n 10 torch.tensor(pydicom.dcmread('ileitis_in_yersiniosis_42_0001.dic').pixel_array)
>>> 207 ms ± 1.59 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: %timeit -n 10 tio.ScalarImage('ileitis_in_yersiniosis_42_0001.dic').data
>>> 691 ms ± 1.29 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Note that I only convert to `torch.tensor` in pydicom for fairness, since `tio.ScalarImage` already returns a tensor.

**Reply by fepegar** [reply]
Given that TorchIO uses SimpleITK to read DICOM, I suggest you open an issue in [their repo](https://github.com/SimpleITK/SimpleITK).

**Reply by vedal** [reply]
@fepegar is it an option to change to pydicom under-the-hood in torchio, as you suggested in your previous answer?

**Reply by fepegar** [reply]
I suppose you can define a function that takes a path, reads the data with pydicom and returns a 4D tensor and a 2D affine, as specified in https://torchio.readthedocs.io/data/image.html#torchio.Image
